### ISSUE TITLE
OpenCV Error: Unsupported format or combination of formats

### ISSUE THREAD

**Issue by rovnanik** [Rolf-Hempel/PlanetarySystemStacker#21]
Hi,
I am having got this error message:

![errorPlanetarySystemStacker_2020-03-12_07-58-41](https://user-images.githubusercontent.com/62076614/76501800-33a10b00-6443-11ea-9250-90326b5f5531.png)

Using PPS 0.7.0, Linux Ubuntu Studio 19.10

Here is the output log:
***********************************************************************
23-00-37.8 Start processing /home/michal/Obrázky/Olympus Workspace/moon
***********************************************************************

           Stacking parameters:                                         | Value             |
           ----------------------------------------------------------------------------------
           Debayering default                                           | Auto detect color |
           Noise level (add Gaussian blur)                              | 7                 |
           Frame stabilization mode                                     | Planet            |
           Percentage of best frames for reference frame computation    | 5                 |
           Alignment box width (pixels)                                 | 48                |
           Max. alignment search width (pixels)                         | 14                |
           Minimum structure                                            | 0.04              |
           Minimum brightness                                           | 10                |
           Percentage of best frames to be stacked                      | 10                |
           Normalize frame brightness                                   | True              |
           Normalization black cut-off                                  | 15                |
          
23-00-37.8 +++ Buffering level is 1 +++
           Number of images: 59, image shape: (3456, 4608, 3)
           Debayer pattern detected automatically: 'Auto detect color'
           Dynamic range of input frames: 8 bit.
           No matching master dark / flat frames found, calibration de-activated
           RAM required (Gbytes): 2.002599936, available: 13.167804416
23-00-37.9 +++ Start ranking images +++
           Index of best frame: 22
23-00-52.8 +++ Initializing frame alignment +++
23-00-52.8 +++ Start aligning all frames +++
OpenCV Error: Unsupported format or combination of formats () in threshold, file /build/opencv-iC2m9y/opencv-3.2.0+dfsg/modules/imgproc/src/thresh.cpp, line 1431
07-59-58.1 Error in aligning frames: /build/opencv-iC2m9y/opencv-3.2.0+dfsg/modules/imgproc/src/thresh.cpp:1431: error: (-210)  in function threshold
, continuing with next job

**Comment by Rolf-Hempel** [comment]
Hi,

Thank you for sending this error report. I have not seen this error before. It seems to occur in line 431 in module "align_frames.py" (method "center_of_gravity").

I don't know how much you know about Python programming. Would you be able to debug the source code with your example and find the reason why the OpenCV routing "threshold" did not work well in this case? Otherwise, could you please give me access to the video file (as short as possible, perhaps a few frames are sufficient?)?

- Rolf

**Comment by Rolf-Hempel** [comment]
Hi,

Something else caught my attention: The path to your example video suggests that you are processing a moon video. Are you sure that the frame stabilization mode "Planet" is appropriate? If this is really a moon video, mode "Surface" most likely is the better choice. For "Planet" the object must be surrounded by black space everywhere in all frames.

Please check if this is really the case.

Regards,
 Rolf

**Comment by rovnanik** [comment]
> Hi,
> 
> Something else caught my attention: The path to your example video suggests that you are processing a moon video. Are you sure that the frame stabilization mode "Planet" is appropriate? If this is really a moon video, mode "Surface" most likely is the better choice. For "Planet" the object must be surrounded by black space everywhere in all frames.
> 
> Please check if this is really the case.
> 
> Regards,
> Rolf

It is not a video, but a consecutive collection of TIFF images of the Moon. As I have only an 400mm objective, the Moon itself is not bigger than a coin and is surrounded by black space all around.

I tried the mode "Surface" at first, it also failed with many warning:
$ python3 planetary_system_stacker.py 
qt5ct: using qt5ct plugin
qt5ct: D-Bus global menu: no
08-59-09.0 Number of threads used by mkl: 2
***********************************************************************
09-16-44.5 Start processing /home/michal/Obrázky/Olympus Workspace/moon
***********************************************************************

           Stacking parameters:                                         | Value             |
           ----------------------------------------------------------------------------------
           Debayering default                                           | Auto detect color |
           Noise level (add Gaussian blur)                              | 7                 |
           Frame stabilization mode                                     | Surface           |
           Automatic frame stabilization                                | True              |
           Stabilization patch size (% of frame size)                   | 33                |
           Stabilization search width (pixels)                          | 34                |
           Percentage of best frames for reference frame computation    | 5                 |
           Alignment box width (pixels)                                 | 48                |
           Max. alignment search width (pixels)                         | 14                |
           Minimum structure                                            | 0.04              |
           Minimum brightness                                           | 10                |
           Percentage of best frames to be stacked                      | 10                |
           Normalize frame brightness                                   | True              |
           Normalization black cut-off                                  | 15                |
          
09-16-44.5 +++ Buffering level is 2 +++
           Number of images: 59, image shape: (3456, 4608, 3)
           Debayer pattern detected automatically: 'Auto detect color'
           Dynamic range of input frames: 8 bit.
           No matching master dark / flat frames found, calibration de-activated
           RAM required (Gbytes): 3.8817792, available: 11.229904896
09-16-44.6 +++ Start ranking images +++
           Index of best frame: 22
09-17-05.6 +++ Initializing frame alignment +++
           Alignment rectangle, computed automatically: 1166<y<2288, 1550<x<3056
09-17-05.7 +++ Start aligning all frames +++
09-17-05.9 Warning: No valid shift computed at frame 20, will try another stabilization patch
           Next alignment rectangle tried: 1166<y<2288, 2303<x<3809
09-17-06.1 Warning: No valid shift computed at frame 20, will try another stabilization patch
           Next alignment rectangle tried: 1727<y<2849, 1550<x<3056
09-17-06.2 Warning: No valid shift computed at frame 20, will try another stabilization patch
           Next alignment rectangle tried: 1727<y<2849, 2303<x<3809
09-17-06.4 Warning: No valid shift computed at frame 20, will try another stabilization patch
           Next alignment rectangle tried: 605<y<1727, 1550<x<3056
09-17-06.6 Warning: No valid shift computed at frame 20, will try another stabilization patch
           Next alignment rectangle tried: 605<y<1727, 2303<x<3809
09-17-06.7 Warning: No valid shift computed at frame 20, will try another stabilization patch
           Next alignment rectangle tried: 1166<y<2288, 797<x<2303
09-17-06.9 Warning: No valid shift computed at frame 20, will try another stabilization patch
           Next alignment rectangle tried: 1727<y<2849, 797<x<2303
09-17-07.1 Warning: No valid shift computed at frame 20, will try another stabilization patch
           Next alignment rectangle tried: 605<y<1727, 797<x<2303
09-17-07.4 Warning: No valid shift computed at frame 19, will try another stabilization patch
           Next alignment rectangle tried: 2288<y<3410, 3056<x<4562
09-17-07.7 Warning: No valid shift computed at frame 19, will try another stabilization patch
           Next alignment rectangle tried: 2288<y<3410, 2303<x<3809
09-17-07.7 Warning: No valid shift computed at frame 21, will try another stabilization patch
           Next alignment rectangle tried: 2288<y<3410, 1550<x<3056
09-17-07.9 Warning: No valid shift computed at frame 20, will try another stabilization patch
           Next alignment rectangle tried: 2288<y<3410, 797<x<2303
09-17-11.3 Warning: No valid shift computed at frame 28, will try another stabilization patch
           Next alignment rectangle tried: 2288<y<3410, 44<x<1550
09-17-12.9 Warning: No valid shift computed at frame 9, will try another stabilization patch
           Next alignment rectangle tried: 1727<y<2849, 3056<x<4562
09-17-13.1 Warning: No valid shift computed at frame 21, will try another stabilization patch
           Next alignment rectangle tried: 1727<y<2849, 44<x<1550
09-17-14.1 Warning: No valid shift computed at frame 14, will try another stabilization patch
           Next alignment rectangle tried: 1166<y<2288, 3056<x<4562
09-17-14.2 Warning: No valid shift computed at frame 20, will try another stabilization patch
           Next alignment rectangle tried: 1166<y<2288, 44<x<1550
09-17-15.0 Warning: No valid shift computed at frame 15, will try another stabilization patch
           Next alignment rectangle tried: 605<y<1727, 3056<x<4562
09-17-15.2 Warning: No valid shift computed at frame 20, will try another stabilization patch
           Next alignment rectangle tried: 605<y<1727, 44<x<1550
09-17-15.6 Warning: No valid shift computed at frame 18, will try another stabilization patch
           Next alignment rectangle tried: 44<y<1166, 3056<x<4562
09-17-16.3 Warning: No valid shift computed at frame 16, will try another stabilization patch
           Next alignment rectangle tried: 44<y<1166, 2303<x<3809
09-17-16.3 Warning: No valid shift computed at frame 21, will try another stabilization patch
           Next alignment rectangle tried: 44<y<1166, 1550<x<3056
09-17-16.5 Warning: No valid shift computed at frame 20, will try another stabilization patch
           Next alignment rectangle tried: 44<y<1166, 797<x<2303
09-17-16.9 Warning: No valid shift computed at frame 18, will try another stabilization patch
           Next alignment rectangle tried: 44<y<1166, 44<x<1550
09-17-17.3 Warning: No valid shift computed at frame 18, will try another stabilization patch
09-17-27.8 Error: Frame stabilization failed at frame 18, continuing with next job. Try a higher value for parameter 'stabilization search width'.

Nothing changed even with a way more higher value for 'stabilization search width'. So I switched to mode "Planet" with the error in subject.
When I look more deeply into warnings of "Surface" mode, it surprises me the frames are not sorted and some of them are listed several times (e. g. frames 18, 20).

I have sent you link to my pictures via email, please try them yourself.

**Comment by Rolf-Hempel** [comment]
Hi,

okay, thanks for sending me your files. I could download them, and I will look into the reason why the "Planet" frame stabilization failed in this case.

In Surface mode, frame stabilization failed because in your frames the moon wanders through the scene (you did not guide the camera). The shift between two consecutive frames is larger than the parameter "stabilization search width". If you raise this value, Surface stabilization mode works without problems. (At least it did so on my computer)

The many warnings you saw above are caused by PSS trying to use different places in the image to do the stabilization. When it has tried all possible locations, finally it gives up. Since in your case the search width was too low, it could not succeed in any location.

Regards,
 Rolf

**Comment by Rolf-Hempel** [comment]
Hi,

Now I checked the "Planet" stabilization on several computers (two with Windows 10, one with Ubuntu 16.04). On all of them your example went through without any warning or error. Also, I checked again the behaviour of the "up" and "down" keys in the configuration dialog. They also worked flawlessly on all machines.

Therefore, the problem must be related to your Python environment or some other incompatibility on your machine. As long as nobody else reports similar problems, I regard this problem as resolved.

All the best,
 Rolf

**Comment by rovnanik** [comment]
Hi,
thank you for your investigation.
Ubuntu 16.04 is quite old so almost everything is different on current
Ubuntu 19.10.
I'll try to figure out what's wrong with my installation.

Regards,
Michal

Dne pá 13. bře 2020 22:09 uživatel Rolf-Hempel <notifications@github.com>
napsal:

**Comment by skolarianer** [comment]
Got the same error even earlier in the process.
PSS 0.7.0, Linux Mint 19.3
Any advice which other log files to check?

```
04-17-44.4 +++ Start ranking images +++
04-17-49.7 Error: /build/opencv-L2vuMj/opencv-3.2.0+dfsg/modules/imgproc/src/thresh.cpp:1431: error: (-210)  in function threshold
, continuing with next job
```

**Comment by Rolf-Hempel** [comment]
Hi,

In the other case I tried to repeat the problem on my computers using the original input data of the user (rovnanik). I could not reproduce the error, neither on Windows, nor on Linux (Ubuntu 16.04).

I have no idea why the error occurs in the OpenCV routine. 

- Rolf

**Comment by skolarianer** [comment]
The same dataset is working fine on Windows for me, too.

I tried OpenCV 4.2.0 instead (installed via pip3) but this breaks PSS completely. Presumably because that changed the version of numpy, too.
`Collecting numpy>=1.11.3 (from opencv-python)`
```
Installing collected packages: numpy, opencv-python
Successfully installed numpy-1.18.2 opencv-python-4.2.0.32
```

Which version of OpenCV do you use on Ubuntu 16.04?

```
$ python3 /opt/PlanetarySystemStacker/Source/planetary_system_stacker.py
Traceback (most recent call last):
  File "/opt/PlanetarySystemStacker/Source/planetary_system_stacker.py", line 40, in <module>
    import skimage
  File "/usr/lib/python3/dist-packages/skimage/__init__.py", line 158, in <module>
    from .util.dtype import *
  File "/usr/lib/python3/dist-packages/skimage/util/__init__.py", line 7, in <module>
    from .arraycrop import crop
  File "/usr/lib/python3/dist-packages/skimage/util/arraycrop.py", line 8, in <module>
    from numpy.lib.arraypad import _validate_lengths
ImportError: cannot import name '_validate_lengths'

```

**Comment by rovnanik** [comment]
Hi,
here is what I've found recently trying debugger:

```
$ python3 -m pdb planetary_system_stacker.py
> /home/michal/work/PlanetarySystemStacker/Source/planetary_system_stacker.py(24)<module>()
-> """
(Pdb) break align_frames.py:431
Breakpoint 1 at /home/michal/work/PlanetarySystemStacker/Source/align_frames.py:431
(Pdb) c
qt5ct: using qt5ct plugin
qt5ct: D-Bus global menu: no
17-54-15.7 Number of threads used by mkl: 2
OpenCV Error: Unsupported format or combination of formats () in threshold, file /build/opencv-iC2m9y/opencv-3.2.0+dfsg/modules/imgproc/src/thresh.cpp, line 1431
```


Here is the definition of the function in opencv-3.2.0+dfsg/modules/imgproc/src/thresh.cpp, line 1431:

`double cv::threshold( InputArray _src, OutputArray _dst, double thresh, double maxval, int type )`

it does not correspond to how the function is called in your code:
`thresh = threshold(frame, frame.max()/2, 1, THRESH_BINARY)[1]`

**Comment by Rolf-Hempel** [comment]
Hi Skolarianer,

Here is the list of library versions used in my Ubuntu Linux 16.04:

![Python-Versions](https://user-images.githubusercontent.com/19596791/78281212-a2412800-751a-11ea-9770-057e1329856e.PNG)

- Rolf

**Comment by Rolf-Hempel** [comment]
Hi Rovnanik,

Why do you think I should use the C++ style of the library call in my Python code? The way I call the routine is consistent with the Python definiton.

- Rolf

**Comment by rovnanik** [comment]
> Hi Rovnanik,
> 
> Why do you think I should use the C++ style of the library call in my Python code? The way I call the routine is consistent with the Python definiton.
> 
> * Rolf

I am no programmer but AFAIK you should use the same number and same types of parameters as in definition ;-)

I'll try to upgrade my opencv-3.2.0 to your version. This one and Pip-18.1 are the only ones older than yours.
![image](https://user-images.githubusercontent.com/62076614/78283487-2517b200-751e-11ea-898e-b76ca3710f96.png)
But these versions are default to Ubuntu 19.10 and probably most users will use them.

**Comment by rovnanik** [comment]
In the new Ubuntu 20.04 (should be released this month) will be python3-opencv-4.2.0+dfsg-5 available.

**Comment by skolarianer** [comment]
I just installed PSS in a `venv` on Linux Mint 19.3. It's working fine in this environment. So, the version of OpenCV shouldn't be the problem.
```
$ pip list
astropy (4.0.1.post1)
cycler (0.10.0)
decorator (4.4.2)
imageio (2.8.0)
intel-openmp (2020.0.133)
kiwisolver (1.2.0)
matplotlib (3.2.1)
mkl (2019.0)
networkx (2.4)
numpy (1.18.2)
opencv-python (4.2.0.32)
Pillow (7.1.1)
pip (9.0.1)
pkg-resources (0.0.0)
psutil (5.7.0)
pyparsing (2.4.6)
PyQt5 (5.14.0)
PyQt5-sip (12.7.1)
python-dateutil (2.8.1)
PyWavelets (1.1.1)
scikit-image (0.16.2)
scipy (1.4.1)
setuptools (39.0.1)
six (1.14.0)
```
